core/src/main/java/org/bitcoinj/coinjoin/CoinJoinClientSession.java

```diff
@@ -0,0 +1,771 @@
+        mixingMasternode = null;
+        pendingDsaRequest = new PendingDsaRequest();
+
+        super.setNull();
+    }
+
+    public CoinJoinClientSession(Wallet mixingWallet) {
+        this.mixingWallet = mixingWallet;
+    }
+
+    // break this up by message:
+//    public void ProcessMessage(Peer from, String msg_type, CDataStream& vRecv, CConnman& connman, bool enable_bip61) {
+//
+//    }
+
+    public void processStatusUpdate(Peer peer, CoinJoinStatusUpdate statusUpdate) {
+        if (mixingMasternode == null)
+            return;
+        if (!mixingMasternode.getService().getAddr().equals(peer.getAddress().getAddr()))
+            return;
+
+        processPoolStateUpdate(statusUpdate);
+    }
+
+    public void processFinalTransaction(Peer peer, CoinJoinFinalTransaction finalTransaction) {
+        if (mixingMasternode == null)
+            return;
+        if (!mixingMasternode.getService().getAddr().equals(peer.getAddress().getAddr()))
+            return;
+
+        if (sessionID.get() != finalTransaction.getMsgSessionID()) {
+            log.info("DSFINALTX -- message doesn't match current CoinJoin session: sessionID: {}  msgSessionID: {}",
+                    sessionID, finalTransaction.getMsgSessionID());
+            return;
+        }
+
+        log.info("DSFINALTX -- txNew {}", finalTransaction.getTransaction()); /* Continued */
+
+        // check to see if input is spent already? (and probably not confirmed)
+        signFinalTransaction(finalTransaction.getTransaction(), peer);
+    }
+
+    public void processComplete(Peer peer, CoinJoinComplete completeMessage) {
+        if (mixingMasternode == null)
+            return;
+        if (!mixingMasternode.getService().getAddr().equals(peer.getAddress().getAddr()))
+            return;
+
+        if (completeMessage.getMsgMessageID().value < MSG_POOL_MIN.value || completeMessage.getMsgMessageID().value > MSG_POOL_MAX.value) {
+            log.info("DSCOMPLETE -- nMsgMessageID is out of bounds: {}", completeMessage.getMsgMessageID());
+            return;
+        }
+
+        if (sessionID.get() != completeMessage.getMsgSessionID()) {
+            log.info("DSCOMPLETE -- message doesn't match current CoinJoin session: nSessionID: {}  nMsgSessionID: {}", sessionID.get(), completeMessage.getMsgSessionID());
+            return;
+        }
+
+        log.info("DSCOMPLETE -- nMsgSessionID {}  nMsgMessageID {} ({})", completeMessage.getMsgSessionID(),
+                completeMessage.getMsgMessageID(), CoinJoin.getMessageByID(completeMessage.getMsgMessageID()));
+
+        completedTransaction(completeMessage.getMsgMessageID());
+    }
+
+    public void unlockCoins() {
+        if (!CoinJoinClientOptions.isEnabled())
+            return;
+
+        while (true) {
+            //TRY_LOCK(mixingWallet.cs_wallet, lockWallet);
+            //if (!lockWallet) {
+            //    UninterruptibleSleep(std::chrono::milliseconds{50});
+            //    continue;
+            //}
```

Comment:
not sure how to translate this yet.

Reply:
`Thread.sleep()` or something like that?

Reply:
Yes, this is true.  From a bigger picture point of view I am wondering which thread would be put to sleep.  if this code is executed in a thread that manages communication with a mixing masternode, then it may cause the peer to disconnect.

This waiting step may actually be one reason why mixing takes a long time:  Waiting for the next block to be mined.

Reply:
I see, makes sense